test/hotspot/jtreg/compiler/c2/irTests/XorINodeIdealizationTests.java

```diff
@@ -217,4 +226,101 @@ public int test16(int x, int y, int z) {
     public int test17(int x) {
         return ~(x + 2022);
     }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    // Checks (c1 ^ c2)  => c3 (constant folded)
+    public int testConstXor() {
+        return CONST_1 ^ CONST_2;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    // Checks (x ^ x)  => c (constant folded)
+    public int testXorSelf(int x) {
+        return x ^ x;
+    }
+
+    private static final boolean CONST_BOOL_1 = RunInfo.getRandom().nextBoolean();
+    private static final boolean CONST_BOOL_2 = RunInfo.getRandom().nextBoolean();
+
+    @Run(test={
+            "testConstXorBool", "testXorSelfBool", "testXorIntAsBool"
+    })
+    public void runBooleanTests() {
+        int c = G.next();
+        int d = G.next();
+
+        assertBooleanResult(true, c, d);
+        assertBooleanResult(false, c, d);
+    }
+
+    @DontCompile
+    public void assertBooleanResult(boolean b, int x, int y) {
+        Asserts.assertEQ(CONST_BOOL_1 ^ CONST_BOOL_2, testConstXorBool());
+        Asserts.assertEQ(false, testXorSelfBool(b));
+        Asserts.assertEQ(true, testXorIntAsBool(x, y));
+    }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    // Checks (c1 ^ c2)  => c3 (constant folded)
+    public boolean testConstXorBool() {
+        return CONST_BOOL_1 ^ CONST_BOOL_2;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    // Checks (x ^ x)  => c (constant folded)
+    public boolean testXorSelfBool(boolean x) {
+        return x ^ x;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    // This test explicitly checks for constant folding over ints representing booleans.
+    // Checks (x ^ y) => z in [0, 1] when x and y are known to be in [0, 1] (constant folded)
+    public boolean testXorIntAsBool(int xi, int yi) {
+        int xor = (xi & 1) ^ (yi & 1);
+        return 0 <= xor && xor <= 1;
+    }
+
+    @Run(test = {
+            "testFoldableXor", "testXorConstRange"
+    })
+    public void runRangeTests() {
+        int a = G.next();
+        int b = G.next();
+        checkXor(a, b);
+
+        for (a = 0; a < 16; a++) {
+            for (b = a; b < 16; b++) {
+                checkXor(a, b);
+            }
+        }
+    }
+
+    @DontCompile
+    public void checkXor(int a, int b) {
+        Asserts.assertEQ(true, testFoldableXor(a, b));
+        Asserts.assertEQ((a & 0b1000) ^ (b & 0b1000), testXorConstRange(a, b));
+    }
+
+    @Test
+    public int testXorConstRange(int x, int y) {
```

Comment:
Should this have an `@IR` test attached, like `testFoldableXor`?

Reply:
This test is really there to do a minimal correctness test of the xor calculation, rather than verifying the IR. The xor is not expected to be eliminated.

Reply:
I've updated that test to make more sense, and include IR checks